core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,477 @@
+              newAddrs.add(addr);
+              if (!subchannels.containsKey(addr)) {
+                List<EquivalentAddressGroup> addrs = new ArrayList<>();
+                addrs.add(new EquivalentAddressGroup(addr));
+                final Subchannel subchannel = helper.createSubchannel(
+                    CreateSubchannelArgs.newBuilder()
+                    .setAddresses(addrs)
+                    .build());
+                subchannels.put(addr, subchannel);
+                subchannel.start(new SubchannelStateListener() {
+                  @Override
+                  public void onSubchannelState(ConnectivityStateInfo stateInfo) {
+                    processSubchannelState(subchannel, stateInfo);
+                  }
+                });
+              }
+            }
+          }
+
+          // remove old subchannels that were not in new address list
+          for (SocketAddress oldAddr : oldAddrs) {
+            if (!newAddrs.contains(oldAddr)) {
+              subchannels.get(oldAddr).shutdown();
+              subchannels.remove(oldAddr);
+            }
+          }
+
+          if (addressIndex.seekTo(previousAddress)) {
+            // start connection attempt at first address
+            if (currentState == CONNECTING || currentState == TRANSIENT_FAILURE) {
+              addressIndex.reset();
+              requestConnection();
+            // TODO: verify desired behavior (ideally, we should restart at first address)
+            // start connection attempt at first address when requested
+            } else if (currentState == IDLE) {
+              addressIndex.reset();
+              SubchannelPicker picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+              updateBalancingState(IDLE, picker);
+            }
+          } else {
+            // start connection attempt at first address
+            if (currentState == CONNECTING || currentState == TRANSIENT_FAILURE) {
+              requestConnection();
+            // start connection attempt at first address when requested
+            } else if (currentState == IDLE || currentState == READY) {
+              SubchannelPicker picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+              updateBalancingState(IDLE, picker);
+            }
+          }
+        }
+      });
+    }
+
+    @Override
+    public void handleNameResolutionError(Status error) {
+      for (Subchannel subchannel : subchannels.values()) {
+        subchannel.shutdown();
+        subchannel = null;
+      }
+      subchannels.clear();
+      // NB(lukaszx0) Whether we should propagate the error unconditionally is arguable. It's fine
+      // for time being.
+      updateBalancingState(TRANSIENT_FAILURE, new Picker(PickResult.withError(error)));
+    }
+
+    void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateInfo) {
```

Comment:
Let's shutdown all other subchannels when one goes ready to avoid noises.
One major issue here is that we should think this callback can be from any subchannel at any time so a lot of noises. Previously in InternalSubchannel, there is activeTransport and pendingTransport which are the ones we only care about. For example, when we receive IDLE here, we know it might be a READY channel goes failure, but we should only care about the one READY channel that we chose before. There might be race condition that other channels also return an IDLE. To distinguish those subchannels, you can either save it to an active/pendingSubchannel, or maybe comparing the current index.

Reply:
That makes sense! I had that as one of my clarifying questions as the proposal says to "deref" the subchannel, but I wasn't sure what this meant. I also waited to make the change in this particular edge case: if READY connections very frequently fail and we need to re-establish a connection, it would be more efficient to not shutdown subchannels since we would need to re-create them. But, if this is not the case, it would make sense to shutdown all connection attempts once one is established to reduce noise.  

Also, there is definitely a lot of unnecessary noise, but I think the behavior should be correct. `L180` allows us to only look at active subchannels by checking the map of active subchannels, which stops updates for subchannels that are in the process of being shutdown. This is tested in the `updateAddresses()` test cases as well, but I'm not positive on how the thread behavior works, so I wanted to run this by you as well.